komga/src/main/kotlin/org/gotson/komga/domain/service/LibraryScanner.kt

```diff
@@ -18,95 +26,238 @@ class LibraryScanner(
   private val seriesRepository: SeriesRepository,
   private val bookRepository: BookRepository,
-  private val bookLifecycle: BookLifecycle,
   private val mediaRepository: MediaRepository,
-  private val seriesLifecycle: SeriesLifecycle
+  private val seriesLifecycle: SeriesLifecycle,
+  private val fileHasher: XXHasher,
+  private val taskReceiver: TaskReceiver,
+  private val bookMetadataRepository: BookMetadataRepository,
+  private val seriesMetadataRepository: SeriesMetadataRepository
 ) {
 
   fun scanRootFolder(library: Library) {
     logger.info { "Updating library: $library" }
     measureTime {
-      val scannedSeries =
-        fileSystemScanner.scanRootFolder(Paths.get(library.root.toURI()), library.scanForceModifiedTime)
-          .map { (series, books) ->
-            series.copy(libraryId = library.id) to books.map { it.copy(libraryId = library.id) }
-          }.toMap()
-
-      // delete series that don't exist anymore
-      if (scannedSeries.isEmpty()) {
-        logger.info { "Scan returned no series, deleting all existing series" }
-        val seriesIds = seriesRepository.findByLibraryId(library.id).map { it.id }
-        seriesLifecycle.deleteMany(seriesIds)
-      } else {
-        scannedSeries.keys.map { it.url }.let { urls ->
-          val series = seriesRepository.findByLibraryIdAndUrlNotIn(library.id, urls)
-          if (series.isNotEmpty()) {
-            logger.info { "Deleting series not on disk anymore: $series" }
-            seriesLifecycle.deleteMany(series.map { it.id })
+      val scannedSeries = fileSystemScanner.scanRootFolder(Paths.get(library.root.toURI()), library.scanForceModifiedTime)
+        .map { (series, books) -> series.copy(libraryId = library.id) to books.map { it.copy(libraryId = library.id) } }
+        .map { (series, books) -> series to books.map { it.copy(fileHash = getBookHash(it)) } }
+        .toMap()
+
+      deleteNoLongerExistingSeries(scannedSeries, library)
+      restoreDeletedSeries(scannedSeries)
+      updateSeriesAndBooks(scannedSeries, library)
+      restoreDeletedBooksAndAddNew(scannedSeries, library)
+    }.also { logger.info { "Library updated in $it" } }
+  }
+
+  private fun getBookHash(book: Book): String {
+    return bookRepository.findByLibraryIdAndUrlIncludeDeleted(book.libraryId, book.url)?.let {
+      if (isModified(it.fileLastModified, book.fileLastModified)) fileHasher.getHash(book.path())
+      else it.fileHash
+    } ?: fileHasher.getHash(book.path())
+  }
+
+  private fun deleteNoLongerExistingSeries(scannedSeries: Map<Series, Collection<Book>>, library: Library) {
+    if (scannedSeries.isEmpty()) {
+      logger.info { "Scan returned no series, mark all existing series as deleted" }
+      seriesRepository.findByLibraryId(library.id)
+        .map { it.id }
+        .let {
+          bookRepository.softDeleteByBookIds(bookRepository.findAllIdBySeriesIds(it))
+          seriesRepository.softDeleteAll(it)
```

Comment:
we should use the Lifecycle services to delete stuff instead of the repos directly. We can probably only soft delete the series, since the Lifecycle service will handle dependencies with books.